torch/utils/data/distributed.py

```diff
@@ -36,23 +45,36 @@ def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True):
         self.num_replicas = num_replicas
         self.rank = rank
         self.epoch = 0
-        self.num_samples = int(math.ceil(len(self.dataset) * 1.0 / self.num_replicas))
+
+        if dataset is None and length is None:
```

Comment:
Doing `DistributedSampler(None, length=5)` is weird, why can't we allow an `int` in place of the dataset?

Reply:
Both approaches work for our use case. I can make the change in a BC way if `dataset` being a `int` and not an instance of `Dataset` feels clearer. If more context is needed, this design was initially discussed [here](https://github.com/pytorch/pytorch/pull/26547#issuecomment-546480531)

Reply:
@apaszke - aside from the reasoning in my comment, branching on the type of an input can also make it easier to make a mistake. Even if we added the option of allowing an int in place of the dataset, it should still make sense to provide the length argument as an additional feature for existing API calls (i.e. those that pass in a map-style dataset).